GVFS/GVFS.UnitTests/Common/TryCreateProductUpgraderTests.cs

```diff
@@ -0,0 +1,163 @@
+﻿using GVFS.Common;
+using GVFS.Common.FileSystem;
+using GVFS.Common.Git;
+using GVFS.Common.NuGetUpgrade;
+using GVFS.Tests.Should;
+using GVFS.UnitTests.Mock.Common;
+using Moq;
+using NUnit.Framework;
+
+namespace GVFS.UnitTests.Common
+{
+    public class TryCreateProductUpgradeTests
+    {
+        private static string defaultUpgradeFeedPackageName = "package";
+        private static string defaultUpgradeFeedUrl = "https://pkgs.dev.azure.com/contoso";
+        private static string defaultRing = "slow";
+
+        private MockTracer tracer;
+        private Mock<PhysicalFileSystem> fileSystemMock;
+        private Mock<ICredentialStore> credentialStoreMock;
+
+        [SetUp]
+        public void Setup()
+        {
+            this.tracer = new MockTracer();
+
+            // It is important that creating a new Upgrader does not
+            // require credentials.  We must be able to create an
+            // upgrader to query / check upgrade preconditions without
+            // requiring authorization.  We create these mocks with
+            // strict behavior to validate methods on them are called
+            // unnecessarily.
+            this.credentialStoreMock = new Mock<ICredentialStore>(MockBehavior.Strict);
+            this.fileSystemMock = new Mock<PhysicalFileSystem>();
+        }

```

Comment:
Do we need a `TearDown` that will call `VerifyAll()` for these mocks?

Reply:
With the `Behavior.Strict`, it will throw an error if there is a call to a method that is not expected (I don't need `VerifyAll` for this). For the `fileSystemMock`, I am using it more to support the test and didn't want to add any verification on it at this time. From another comment, there is an existing call in one of the GitHubUpgrader code paths that might not be necessary, but we still have to account for it now....

Reply:
> For the fileSystemMock, I am using it more to support the test and didn't want to add any verification on it at this time.

As discussed offline for a sanity check you could consider mocking `CreateDirectory` and setting `Behavior.Strict` so that we can catch if new dependencies are introduced.

Reply:
Done!